### PR TITLE
Use GL context to draw in a thread save way

### PR DESCRIPTION
**Pull request by PhilippvK**
Feel free to ask me anything if you have questions. Spent multiple days (and nights) figuring this out, getting it work on every operating system (inside Containers).

I probably won't answer today, need some rest first.

### CODE REVIEW

**Comment by PhilippvK** [comment]
Will need some time to rebase this branch to your latest changes

**Comment by alxhoff** [comment]
ok. i'll wait for you to do it and then merge it. What exactly does this do?

**Comment by PhilippvK** [comment]
I will annotate some sourcecode for you when the conflicts are resolved to answer your question.

**Comment by alxhoff** [comment]
Awesome. First a coffee and then i'll check it out. Cheers!

**Comment by PhilippvK** [comment]
No hurries I probably won’t tackle it before lunch. Sorry!

**Comment by alxhoff** [comment]
No worries. Have plenty of other stuff to do.

File: lib/Gfx/TUM_Draw.c [hunk 1]
```diff
@@ -168,6 +168,7 @@ const int screen_width = SCREEN_WIDTH;
 
 SDL_Window *window = NULL;
 SDL_Renderer *renderer = NULL;
+SDL_GLContext context = NULL;
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:171]: A GL Context is required make SDL2 drawing possible among multiple threads on every platform.

Only one thread can own the current context and draw to the window.

File: lib/Gfx/TUM_Draw.c [hunk 2]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
-    }
-
-	window = SDL_CreateWindow(WINDOW_TITLE, SDL_WINDOWPOS_CENTERED,
-				  SDL_WINDOWPOS_CENTERED, screen_width,
-				  screen_height, SDL_WINDOW_SHOWN);
-
-	if (!window) {
-        PRINT_SDL_ERROR("failed to create %d x %d window '%s'", screen_width, screen_height, WINDOW_TITLE);
-        goto err_window;
+	static char firstCall = 1; // Used to transfer the Thread Context
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: While `SDL_Init ` always has to be called from the Thread 0 running `main()`, at the updates of the screen have to be done in another thread. I solved this by claiming the context and creating the renderer in the second call of the function. I am sure there are other alternatives like having 2 separate functions which are less hacky.

File: lib/Gfx/TUM_Draw.c [hunk 3]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+#warning "HOST_OS undefined! Assuming 'linux'..."
+#elif HOST_OS != linux
+		setenv("LIBGL_ALWAYS_INDIRECT","1", 1); // speed up drawings a little bit
+		setenv("SDL_VIDEO_X11_VISUALID", "", 1); // required on windows and macos
+#elif HOST_OS == linux
+		// nothing
+#else
+#error "Unexpected value of HOST_OS!"
+#endif /* HOST_OS */
+#endif /* DOCKER */
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:622]: Just ignore this. (Exporting some environment variables is required on some systems when drawing via X11)

File: lib/Gfx/TUM_Draw.c [hunk 4]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+		free(buffer);
+
+		if (!font) {
+			PRINT_TTF_ERROR("Opening font @ '%s' failed", buffer);
+			goto err_open_font;
+		}
+
+		window = SDL_CreateWindow(WINDOW_TITLE, SDL_WINDOWPOS_CENTERED,
+					SDL_WINDOWPOS_CENTERED, screen_width,
+					screen_height, SDL_WINDOW_OPENGL);
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: `SDL_WINDOW_OPENGL` flag is required to use GL contexts.

File: lib/Gfx/TUM_Draw.c [hunk 5]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+
+		if (!window) {
+			PRINT_SDL_ERROR("failed to create %d x %d window '%s'", screen_width, screen_height, WINDOW_TITLE);
+			goto err_window;
+		}
+
+		context = SDL_GL_CreateContext(window);
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: Create Context and check for errors

File: lib/Gfx/TUM_Draw.c [hunk 6]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+
+		if (SDL_GL_MakeCurrent(window, context) < 0) {
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: Main Thread claims context and releases it afterwards.

I think I can get rid of both calls as the there is no drawing happening but I left it in to have the ability to create a renderer and add initial draw statements here which will appear before the Sheduler is running.

File: lib/Gfx/TUM_Draw.c [hunk 7]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+			PRINT_SDL_ERROR("Claiming current context failed", buffer);
+			goto err_make_current;
+		}
+
+		if (SDL_GL_MakeCurrent(window, NULL) < 0) {
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: `NULL` releases the current context. Now its is free to be claimed by another thread.

File: lib/Gfx/TUM_Draw.c [hunk 8]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+			goto err_make_current;
+		}
+
+		atexit(SDL_Quit);
+
+		return 0;
+
+	} else { // Should be called from the Drawing Thread
+
+		if (SDL_GL_MakeCurrent(window, context) < 0) { // Claim GL Context
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: Now the Drawing Thread (`vSwapBuffers`) can claim the context.

File: lib/Gfx/TUM_Draw.c [hunk 9]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+			PRINT_SDL_ERROR("Releasing current context failed", buffer);
+			goto err_make_current;
+		}
+
+		renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED |
+							SDL_RENDERER_TARGETTEXTURE | SDL_RENDERER_PRESENTVSYNC);
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: After the context is owned by `vSwapBuffers`, a renderer can be created.

File: lib/Gfx/TUM_Draw.c [hunk 10]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+			PRINT_SDL_ERROR("Claiming current context failed", buffer);
+			goto err_make_current;
+		}
+
+		if (SDL_GL_MakeCurrent(window, NULL) < 0) {
+			PRINT_SDL_ERROR("Releasing current context failed", buffer);
+			goto err_make_current;
+		}
+
+		atexit(SDL_Quit);
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: `atexit` only required in first call

File: lib/Gfx/TUM_Draw.c [hunk 11]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
-	atexit(SDL_Quit);
-
-    return 0;
-
 err_renderer:
-    SDL_DestroyWindow(window);
+	SDL_DestroyWindow(window);
+err_make_current:
+	SDL_GL_DeleteContext(context);
+err_create_context:
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:681]: Proper Cleanup of context.

File: lib/Gfx/TUM_Draw.c [hunk 12]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
 err_open_font:
 err_font_loc:
-    TTF_Quit();
+	TTF_Quit();
 err_ttf:
-    SDL_Quit();
+	SDL_Quit();
 err_sdl:
-    return -1;
+	return -1;
```
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:690]: You have screwed up the spaces and tabs somehow in a earlier commit. I fixed them here, as I had one indention level more due to the additional if statement.

**Comment by PhilippvK** [comment]
Comments are there, but i still have to test the result. Please wait until done.

**Comment by PhilippvK** [comment]
Yeah the check is failing, I will fix that..

**Comment by PhilippvK** [comment]
I last fix is coming soon...

File: lib/Gfx/TUM_Draw.c [hunk 13]
```diff
@@ -606,66 +607,109 @@ char *tumGetErrorMessage(void)
...
+		if(!buffer){
+			PRINT_TTF_ERROR("Prepending font path failed");
+			goto err_font_loc;
+		}
+
+		font = TTF_OpenFont(buffer, DEFAULT_FONT_SIZE);
+		free(buffer);
+
+		if (!font) {
+			PRINT_TTF_ERROR("Opening font @ '%s' failed", buffer);
```
Comment (alxhoff) [inline lib/Gfx/TUM_Draw.c]: `buffer` here should be replaced with `FONT_LOCATION`. I just realized.

**Review by alxhoff — CHANGES_REQUESTED** [review]

**Comment by PhilippvK** [comment]
I think I am finally done.
I can rebase the commit when you are done reviewing, as I do not want to dismiss the comments now.

File: src/main.c [hunk 14]
```diff
@@ -180,6 +180,8 @@ void vSwapBuffers(void *pvParameters)
 	xLastWakeTime = xTaskGetTickCount();
 	const TickType_t frameratePeriod = 20;
 
+	vInitDrawing(NULL); // Setup Rendering handle with correct GL context
```
Comment (alxhoff) [inline src/main.c]: Without digging into it, should this be NULL?

**Review by alxhoff — COMMENTED** [review]
Drawing init, should it be passed NULL when called from frame swap?

**Review by PhilippvK — COMMENTED** [review]
> Drawing init, should it be passed NULL when called from frame swap?

It is only used on the first call, so argument is not relevant.

I think having 2 separate functions instead of one would make more sense... (arrrggg)

Do you have recommendations for the naming of the 2nd function used to claim the context and create the renderer?

File: lib/Gfx/TUM_Draw.c (hunk 2 above)
Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c]: Ahhh, makes sense. I think two functions would be a good idea. Maybe something like "bindDrawingToTask"?

File: lib/Gfx/TUM_Draw.c (hunk 6 above)
Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c]: Sounds like a good idea to me. I also have a tendency to leave code that will make future additions easier. Probably bad style but meh.

File: lib/Gfx/TUM_Draw.c (hunk 7 above)
Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c]: Did you find a good site on this context stuff? or is it just in docs? Would be interested in giving it a proper read.

**Review by alxhoff — APPROVED** [review]

**Comment by alxhoff** [comment]
will merge once I am done with my current stuff, working on #20.

**Comment by PhilippvK** [comment]
Its 2 weeks ago since i read about it but i will try to find something

**Comment by PhilippvK** [comment]
Should i restructure everything to be 2 separate function before you merge?

**Comment by alxhoff** [comment]
> Should i restructure everything to be 2 separate function before you merge?

yer not a bad idea!

**Comment by PhilippvK** [comment]
https://discourse.libsdl.org/t/sdl-opengl-context-and-threads/17813/5

**Comment by alxhoff** [comment]
thanks!

**Comment by PhilippvK** [comment]
This should be it for now....

File: lib/Gfx/TUM_Draw.c [hunk 15]
```diff
@@ -604,68 +605,121 @@ char *tumGetErrorMessage(void)
...
+	}
 
 	atexit(SDL_Quit);
 
-    return 0;
+	return 0;
 
 err_renderer:
-    SDL_DestroyWindow(window);
+	SDL_DestroyWindow(window);
```
Comment (alxhoff) [inline lib/Gfx/TUM_Draw.c:678]: I think the error handling here is a bit mixed up. I will modify it and you can review.

**Review by alxhoff — COMMENTED** [review]
This appears to of fixed #5 :D

File: lib/Gfx/TUM_Draw.c (hunk 15 above)
Comment (PhilippvK) [reply lib/Gfx/TUM_Draw.c:678]: > I think the error handling here is a bit mixed up. I will modify it and you can review.

You probably habe to fix it in both functions. Sorry.

Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c:678]: other one looks good! So glad #5 is fixed. Great work!

**Pull request merged**